rfc/rfc-48/rfc-48.md

```diff
@@ -0,0 +1,171 @@
+s3://<table_storage_bucket>/0bfb3d6e/<hudi_table_name>/a62aa56b-d55e-4a2b-88a6-d603ef26775c-0_8-11-50_20220301005056692.parquet
+...
+```
+
+The original table path will continue to store the `metadata folder` and `partition metadata` files:
+```
+s3://<table_bucket>/<hudi_table_name>/.hoodie/...
+s3://<table_bucket>/<hudi_table_name>/country=usa/.hoodie_partition_metadata
+s3://<table_bucket>/<hudi_table_name>/country=india/.hoodie_partition_metadata
+s3://<table_bucket>/<hudi_table_name>/country=uk/.hoodie_partition_metadata
+...
+```
+
+#### Hashing
+
+To generate the prefixes we can use `Murmur 32 bit` hash on the `File Names`, which is known for being fast and provides
+good distribution guarantees. We might have to further do bucketing and re-hash it to reduce the number of possible hashes
+from 2^32 to a slightly lower number, as it may be overkill to have that many unique hashes, which might result in scenarios
+where each file is under a different prefix.
+
+### Maintain mapping to files
+
+In [RFC-15](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=147427331), we introduced an internal
+Metadata Table with a `files` partition that maintains mapping from partitions to list of files in the partition stored
+under `Table Path`. This mapping is kept up to date, as operations are performed on the original table. We will leverage
+the same to now maintain mappings to files stored at `Table Storage Path` under different prefixes.
+
+Here are some of the design considerations:
+
+1. Metadata table is a pre-requisite and will be automatically enabled if object storage layout is enabled.
+
+2. The object storage layout cannot be enabled on an existing table that is already bootstrapped with Hive storage layout.
+To switch to object storage layout, the table will need to be re-bootstrapped with the new layout.
+
+3. The Instant metadata (`HoodieCommitMetadata`,`HoodieCleanMetadata` etc.) will always act as the source of file listing
+for metadata table to be populated.
+
+4. `HoodieCommitMetadata` currently stores `file name` instead of complete `file path`. We will have to modify commit
+metadata to store the complete file path instead of just file name, as the files are now distributed across several random
+prefix paths instead of a derivable table/partition path.
+
+5. If there is an error reading from Metadata table, we will not fall back listing from file system.
+
+6. In case of metadata table getting corrupted or lost, we need to have a solution here to reconstruct metadata table
+from the files which distributed using object storage layout. We will likely have to implement a file system listing
+logic, that can get all the partition to files mapping by listing all the prefixes under the `Table Storage Path`.
+Following the folder structure of adding table name/partitions under the prefix will help in getting the listing and
+identifying the table/partition they belong to.
+
+### Query Side Integration
```

Comment:
Thinking out loud, the main thing here is - even files under the same partition will not be spread across different prefixes. I think all the fileSystemView methods that fetch file slices for a given partition, can be made to work as-is. 
Something to watch for is code that assumes all files under a logical partition now lives the same folder i.e using `partitionPath + fileName` as the path. Using `BaseFile#fullPath`  consistently should help address this